graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+
+                input_df = input_df.withColumnRenamed(column, new_column)
+                self.column_substitutions[column] = new_column
+
+        return input_df
+
+    def _write_df(
+        self,
+        input_df: DataFrame,
+        full_output_path: str,
+        out_format: str = "parquet",
+        num_files: Optional[int] = None,
+    ) -> Sequence[str]:
+        """Write a DataFrame to S3 or local storage, in the requested format (csv or parquet).
+
+        If self.num_output_files is set will always coalesce files to that number before writing.
+
+
+        Parameters
+        ----------
+        input_df : DataFrame
+            Spark DataFrame to be written.
+            Could be single column vector or multiple columns.
+        full_output_path : str
+            S3 URI or local path, under which the files will be written.
+            We append the format name (csv/parquet) to the path before writing.
+        out_format : str, optional
+            "csv" or "parquet", by default "parquet"
+        num_files : Optional[int], optional
+             Number of partitions the file will be written with.
+            If None or -1 we let Spark choose the appropriate number. Note that choosing
+            a small number of partitions can have an adverse effect on performance.
+            Overrides value of self.num_output_files.
+
+        Returns
+        -------
+        Sequence[str]
+            A (sorted) list of S3 URIs or local file paths that
+            were written by Spark under the output_path provided.
+
+        Raises
+        ------
+        NotImplementedError
+            If an output format other than "csv" or "parquet" is requested.
+        """
+        if self.filesystem_type == "s3":
+            output_bucket = full_output_path.split("/")[2]
+            output_prefix = full_output_path.split("/", 3)[3]
+        else:
+            output_bucket = ""
+            output_prefix = full_output_path
+
+        if num_files and num_files != -1:
+            input_df = input_df.coalesce(num_files)
+        elif self.num_output_files:
+            input_df = input_df.coalesce(self.num_output_files)
+
+        if out_format == "parquet":
+            # Write to parquet
+            input_df = self._replace_special_chars_in_cols(input_df)
+            input_df.write.mode("overwrite").parquet(os.path.join(full_output_path, "parquet"))
+            prefix_with_format = os.path.join(output_prefix, "parquet")
+        elif out_format == "csv":
+            # TODO: Upstream perhaps it's a good idea for all transformers to guarantee
+            # same output type, e.g. list of floats
+            # Write to CSV
+            def csv_row(data: Row):
+                row_dict = data.asDict()
+                row_vals = list(row_dict.values())
+                if len(row_vals) > 1:
+                    # In case there's more than one column
+                    return ",".join(str(d) for d in row_vals)
```

Comment:
I cannot think of a case where that would happen but let me add an assertion to make sure.

Reply:
For example, the data is string and contains ','. 
Can this happen?

Reply:
In the current implementation no, as we only support numerical output. 

If we add text pass-through (pass text column unchanged to be used by BERT downstream) as GConstruct does, it would still be impossible to happen, because this check requires the incoming DF to have more than 1 column, which would not happen for feature DFs (we write one single-column DF per feature).

That said, I've added an assertion to future-proof this. When we add pass-through text features we'll revisit.